### ISSUE TITLE
Oracle12 visitor - Combination of limit and lock is not supported

### ISSUE THREAD

**Issue by akostadinov** [rsim/oracle-enhanced#2237]
`Oracle12` Arel visitor does not handle sub-queries with `LIMIT` properly and this makes locking a row impossible. While it works in normal queries, in sub-queries that doesn't seem to be the case.

### Steps to reproduce

```
# just a minimal project `rails new --database=oracle
git clone https://github.com/akostadinov/rails-ora-02014.git
cd rails-ora-02014
vi config/database.yml # set your DB connection details
rails c
> a = Article.create!(title: "test", body: "test body")
> a.with_lock {}
```

### Expected behavior
```
  Article Load (1.8ms)  SELECT "ARTICLES".* FROM "ARTICLES" WHERE "ARTICLES"."ID" = :a1 AND ROWNUM <= :a2  FOR UPDATE  [["id", 23], ["LIMIT", 1]]
=> nil
```

### Actual behavior
```
Traceback (most recent call last):
/home/avalon/.asdf/installs/ruby/2.7.3/lib/ruby/gems/2.7.0/gems/activerecord-oracle_enhanced-adapter-7.0.0/lib/arel/visitors/oracle12.rb:10:in `visit_Arel_Nodes_SelectStatement': Combination of limit and lock is not supported. Because generated SQL statements (ArgumentError)
`SELECT FOR UPDATE and FETCH FIRST n ROWS` generates ORA-02014.
```

### Possible workaround - use `Arel::Visitors::Oracle`

`Arel::Visitors::Oracle`  works fine though because it always translates `LIMIT` statements to `ROWNUM`.

The example above properly works, if we add to `application.rb`:

```
ActiveSupport.on_load(:active_record) do
  ActiveRecord::ConnectionAdapters::OracleEnhancedAdapter.class_eval do
    # Use old visitor for Oracle 12c database
    self.use_old_oracle_visitor = true
  end
end
```

### System configuration
**Oracle Enhanced Adapter version**
7.0.0

**Rails version**:
Rails 7.0.1

**Ruby version**:
ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-linux]

**Oracle version**:
19.3.0

**Comment by akostadinov** [comment]
Seems to be related also to #1070

**Comment by ajaya** [comment]
@yahonda 

I see a possible solution from 2016. 
https://github.com/rsim/oracle-enhanced/issues/1066

Is this still the current approach?

AJ

**Comment by joerixaop** [comment]
One possible approach is using a query without limits. Assuming you're using where on a unique value it shouldn't really make a difference in amount of objects returned (and if you're not doing it on a unique value then the setting a limit of 1 is not going to reliably work anyway).

```
a = Article.create!(title: "test", body: "test body")
a = Article.lock(true).where(id: a.id).to_a[0]
...
a.save! # Make sure to unlock the row (if you are running in a transaction then ending the transaction should work too I think)
```

**Comment by akostadinov** [comment]
@joerixaop , thank you for the suggested workaround. It seems possible to workaround this under the hood like (on Rails 6.1.x):
```rb
    ActiveRecord::Relation.prepend(Module.new do
      def find_one(id)
        if ActiveRecord::Base === id
          raise ArgumentError, <<-MSG.squish
            You are passing an instance of ActiveRecord::Base to `find`.
            Please pass the id of the object by calling `.id`.
          MSG
        end

        relation = where(primary_key => id)
        record = relation.to_a.first # this is the only change from the original method

        raise_record_not_found_exception!(id, 0, 1) unless record

        record
      end
    end)
```

I think it is alright to patch all `#find_one`  invocations, because finding by primary key should always find only one element and the database should optimize automatically without needing a hint with `FETCH FIRST`.

FYI my original workaround to use `use_old_oracle_visitor` is not viable on Rails 6.1 because some queries are messed up due to inserting `ROWNUM` field. It appears to be caused by https://github.com/rails/rails/commit/2ad2425fd3b3a58e937f97b1fa05306c4d3166ae#diff-f35ba011d4738058fa067289f7585a0224ef8ab22efe7a7ecdb07c2fed6940eaR419-R420

But since I don't expect anybody to be using an old oracle version anymore, I don't see much sense to file a separate bug about it. Much better would be to fix the new visitor somehow.

**Comment by akostadinov** [comment]
The above patch perhaps needs to also take into account composite primary keys, just putting a note here as I don't have time right now to investigate what is the latest code state.

**Comment by yahonda** [comment]
Heads-up: this issue is referenced as `Closes #2237` in #2573, so it will close automatically when that PR merges.

Some background on why #2573 covers this issue's scope:

- The canonical reproducer in this issue (`a.with_lock {}`) emits `WHERE id = :a1 AND ROWNUM <= :a2 ... FOR UPDATE` after #2573, which is what the issue's "Expected behavior" section asks for.
- The workarounds proposed in this thread — `use_old_oracle_visitor = true` (which @akostadinov noted breaks on Rails 6.1+ due to preloader interaction), the per-call `.to_a[0]` rewrite, and the `find_one` monkey-patch (no composite-PK support) — are no longer needed once #2573 is in: the fix lives in the Oracle12 visitor itself and is scoped to the limit+lock combination, so other queries are unaffected.
- The compound shape — combining `.lock` with `ORDER BY` / `OFFSET` / `GROUP BY` / `HAVING` / `DISTINCT` — still fails. That is unavoidable at the adapter level: Oracle Database itself raises `ORA-02014` for `FOR UPDATE` inside a `ROWNUM` subquery (or under `FETCH FIRST`) regardless of how the SQL is shaped. After #2573 the failure surfaces as an ordinary `ActiveRecord::StatementInvalid` from the database, not a Ruby `ArgumentError` raised from inside an internal Arel visitor — so application code can `rescue ActiveRecord::StatementInvalid` and the error number points users at Oracle's documentation rather than at adapter internals.

See the PR description on #2573 for the full behavior matrix.
